pkg/sql/catalog/schemaexpr/expr.go

```diff
@@ -74,29 +75,27 @@ func DequalifyAndTypeCheckExpr(
 // tree.TypedExpr would be dangerous. It contains dummyColumns which do not
 // support evaluation and are not useful outside the context of type-checking
 // the expression.
-func DequalifyAndValidateExpr(
+func DequalifyAndValidateExprImpl(
 	ctx context.Context,
-	desc catalog.TableDescriptor,
 	expr tree.Expr,
 	typ *types.T,
 	context string,
 	semaCtx *tree.SemaContext,
 	maxVolatility volatility.V,
 	tn *tree.TableName,
+	getAllNonDropColumnsFn func() colinfo.ResultColumns,
+	columnLookupByNameFn func(columnName tree.Name) (exists bool, accessible bool, id catid.ColumnID, typ *types.T),
 ) (string, *types.T, catalog.TableColSet, error) {
 	var colIDs catalog.TableColSet
-	nonDropColumns := desc.NonDropColumns()
-	sourceInfo := colinfo.NewSourceInfoForSingleTable(
-		*tn, colinfo.ResultColumnsFromColumns(desc.GetID(), nonDropColumns),
-	)
+	sourceInfo := colinfo.NewSourceInfoForSingleTable(*tn, getAllNonDropColumnsFn())
```

Comment:
since we always call `getAllNonDropColumnsFn()`, how about we just pass in a `colinfo.ResultColumns`?

Reply:
I will wait a bit on this part bc we might want to use an interface (or something) to make this cleaner. But that will be in a different PR!